modules/webbrowser/src/processors/processorcefsynchronizer.cpp

```diff
@@ -0,0 +1,101 @@
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2018-2019 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/webbrowser/processors/processorcefsynchronizer.h>
+#include <modules/webbrowser/webbrowsermodule.h>
+
+#include <inviwo/core/util/stringconversion.h>
+#include <inviwo/core/processors/progressbarowner.h>
+
+#include <warn/push>
+#include <warn/ignore/all>
+#include "include/cef_parser.h"
+#include <warn/pop>
+
+namespace inviwo {
+
+void ProcessorCefSynchronizer::OnLoadEnd(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame,
+                                         int /*httpStatusCode*/) {
+    // synchronize all properties
+    // Ok to send javascript commands when frame loaded
+    for (auto& observer : progressObservers_) {
+        observer.second.setFrame(frame);
+    }
+}
+
+bool ProcessorCefSynchronizer::OnQuery(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame,
+                                       int64, const CefString& request, bool,
+                                       CefRefPtr<Callback> callback) {
+
+    const std::string& requestStr = request;
+    // Assume format "id":"htmlId"
+    auto j = json::parse(requestStr);
+
+    try {
+        auto command = j.at("command").get<std::string>();
+        auto subscribeProgress = std::string("processor.progress.subscribe");
+        auto network = InviwoApplication::getPtr()->getProcessorNetwork();
+        if (!command.compare(0, subscribeProgress.size(), subscribeProgress)) {
+            auto p = j.at("path").get<std::string>();
```

Comment:
is this code used in more than one place? Looks like it could be moved to utils...

Reply:
I guess so, but for now it is only used in two places. Let's reconsider if it changes